### ISSUE TITLE
Include source build instructions

### ISSUE THREAD

**Issue by alerque** [tobias-klein/ezra-project#1]
I'm trying to setup packages for this on Arch Linux (tagged release + git HEAD), but I am not super familiar with Electron. Could you include a brief section in the README describing what commands should be run to build this from source rather from the prebuilt release binaries?

**Comment by tobias-klein** [comment]
Hi Caleb,

Thanks for your interest and initiative! I actually got the same 
question on the sword mailinglist just recently.
It's exciting to see that you want to setup packages for Arch Linux. I'm 
not familiar with that distribution. I'm personally using Ubuntu/Debian.

I just added build/release instructions for Linux (based on Ubuntu/Debian):
https://github.com/tobias-klein/ezra-project/blob/master/README.md#building-ezra-project

Note that for some reason the previously published release 0.7.1 has 
some compilation issues with Electron right now for some reason.
HEAD should happily compile, though.

Please let me know how it goes and feel free to contact me for any 
additional questions.

Happy Easter!

Best regards,
Tobias

**Comment by alerque** [comment]
Thanks, that helps. At the moment those instructions don't work for me. Here's the error I get when running `npm install`:

```
> ezra-sword-interface@0.4.0 install /home/caleb/projects/aur/ezra-project-git/src/ezra-project/node_modules/ezra-sword-interface
> node-gyp rebuild

make: Entering directory '/home/caleb/projects/aur/ezra-project-git/src/ezra-project/node_modules/ezra-sword-interface/build'
  CC(target) Release/obj.target/nothing/../node-addon-api/src/nothing.o
  AR(target) Release/obj.target/../node-addon-api/src/nothing.a
  COPY Release/nothing.a
  CXX(target) Release/obj.target/ezra_sword_interface/ezra_sword_interface.o
../ezra_sword_interface.cpp: In constructor ‘EzraSwordInterface::EzraSwordInterface()’:
../ezra_sword_interface.cpp:93:26: warning: ‘sword::SWConfig::Sections’ is deprecated [-Wdeprecated-declarations]
         this->_swConfig->Sections.clear();
                          ^~~~~~~~
In file included from /usr/include/sword/swmodule.h:30,
                 from ../ezra_sword_interface.cpp:40:
/usr/include/sword/swconfig.h:125:26: note: declared here
  SWDEPRECATED SectionMap Sections;
                          ^~~~~~~~
../ezra_sword_interface.cpp:93:26: warning: ‘sword::SWConfig::Sections’ is deprecated [-Wdeprecated-declarations]
         this->_swConfig->Sections.clear();
                          ^~~~~~~~
In file included from /usr/include/sword/swmodule.h:30,
                 from ../ezra_sword_interface.cpp:40:
/usr/include/sword/swconfig.h:125:26: note: declared here
  SWDEPRECATED SectionMap Sections;
                          ^~~~~~~~
../ezra_sword_interface.cpp:93:26: warning: ‘sword::SWConfig::Sections’ is deprecated [-Wdeprecated-declarations]
         this->_swConfig->Sections.clear();
                          ^~~~~~~~
In file included from /usr/include/sword/swmodule.h:30,
                 from ../ezra_sword_interface.cpp:40:
/usr/include/sword/swconfig.h:125:26: note: declared here
  SWDEPRECATED SectionMap Sections;
                          ^~~~~~~~
../ezra_sword_interface.cpp:95:31: warning: ‘void sword::SWConfig::Save()’ is deprecated [-Wdeprecated-declarations]
         this->_swConfig->Save();
                               ^
In file included from /usr/include/sword/swmodule.h:30,
                 from ../ezra_sword_interface.cpp:40:
/usr/include/sword/swconfig.h:140:20: note: declared here
  SWDEPRECATED void Save() { save(); }
                    ^~~~
../ezra_sword_interface.cpp: In member function ‘int EzraSwordInterface::installModule(std::__cxx11::string, std::__cxx11::string)’:
../ezra_sword_interface.cpp:486:30: warning: ‘virtual signed char sword::SWMgr::Load()’ is deprecated [-Wdeprecated-declarations]
             this->_mgr->Load();
                              ^
In file included from ../ezra_sword_interface.cpp:41:
/usr/include/sword/swmgr.h:407:35: note: declared here
  SWDEPRECATED virtual signed char Load() { return load(); }
                                   ^~~~
../ezra_sword_interface.cpp: In member function ‘int EzraSwordInterface::uninstallModule(std::__cxx11::string)’:
../ezra_sword_interface.cpp:502:26: warning: ‘virtual signed char sword::SWMgr::Load()’ is deprecated [-Wdeprecated-declarations]
         this->_mgr->Load();
                          ^
In file included from ../ezra_sword_interface.cpp:41:
/usr/include/sword/swmgr.h:407:35: note: declared here
  SWDEPRECATED virtual signed char Load() { return load(); }
                                   ^~~~
  CXX(target) Release/obj.target/ezra_sword_interface/js_ezra_sword_interface.o
  CXX(target) Release/obj.target/ezra_sword_interface/binding.o
  SOLINK_MODULE(target) Release/obj.target/ezra_sword_interface.node
g++: error: /usr/lib/x86_64-linux-gnu/libcurl.so.4: No such file or directory
make: *** [ezra_sword_interface.target.mk:140: Release/obj.target/ezra_sword_interface.node] Error 1
make: Leaving directory '/home/caleb/projects/aur/ezra-project-git/src/ezra-project/node_modules/ezra-sword-interface/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:193:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:255:12)
gyp ERR! System Linux 5.0.7-arch1-1-ARCH
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/caleb/projects/aur/ezra-project-git/src/ezra-project/node_modules/ezra-sword-interface
gyp ERR! node -v v11.14.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
npm WARN ezra-project@0.8.0-Dev No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! ezra-sword-interface@0.4.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the ezra-sword-interface@0.4.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/caleb/.npm/_logs/2019-04-19T09_40_45_176Z-debug.log

```

The full log file: [2019-04-19T09_32_48_753Z-debug.log](https://github.com/tobias-klein/ezra-project/files/3097706/2019-04-19T09_32_48_753Z-debug.log)

At least the warning about the malformed package.json file is something on your end, but I'm not sure what the curl error is about.

**Comment by tobias-klein** [comment]
Looks like the curl library is not available on your system. Could you install it and try again? On Ubuntu 18.04. the libsword package depends on libcurl3-gnutls. But locally I also have libcurl4 installed. Could you first try with libcurl3 and if that does not work libcurl4? I may have to update the build instructions afterwards accordingly.

**Comment by alerque** [comment]
I have `curl-7.64.1` installed already, which has SO  version 4. The latest SVN version of the SWORD libraries compile fine.

**Comment by tobias-klein** [comment]
Ok. I think I found the root-cause to this build issue. The [binding.gyp](https://github.com/tobias-klein/ezra-sword-interface/blob/master/binding.gyp) file of ezra-sword-interface links directly against /usr/lib/x86_64-linux-gnu/libcurl.so.4. I need to figure out how to make this more generic.

**Comment by tobias-klein** [comment]
In the meantime you can try to adjust the binding.gyp file to make it work with your local curl library. You should find the binding.gyp file under `<project-dir>/node_modules/ezra-sword-interface/binding.gyp`. After an adjustment you can just execute the command `npm run rebuild-linux`.

**Comment by alerque** [comment]
I see how I could probably do that on a manual install, but that makes it virtually impossible to package  because there is a race condition. The file I would need to patch does not even exist at the time `npm install` is  run. The procedure would be to run `npm install`, wait for it to die, patch the file inside the downloaded module, then manually rerun the init for that module, then re-run the main npm install ignoring that module. Inside a packaging script, commands are not really allowed to fail. Not being able to run `npm install` successfully is pretty much a no-go for packaging purposes.

**Comment by tobias-klein** [comment]
That's correct! I fully agree with you. I'll be looking into making the binding.gyp file more generic.

**Comment by tobias-klein** [comment]
The pkg-config change is the right approach to this, thanks for the patch.

To make it work on my system I had to first install libcurl4-gnutls-dev package. Before I had that package installed, no *.pc file for pkg-config was available for curl.

Secondly I had to change the pkg-config command from `pkg-config --libs curl` to `pkg-config --libs libcurl`(added _lib_ before curl). How does it look on your system? Does `pkg-config --libs libcurl` also produce a valid output? With `--libs curl`I get this output and the node module does not build:

```
$ pkg-config --libs curl
Package curl was not found in the pkg-config search path.
Perhaps you should add the directory containing `curl.pc'
to the PKG_CONFIG_PATH environment variable
No package 'curl' found
```

**Comment by alerque** [comment]
@tobias-klein `pkg-config --libs libcurl` is correct on my system as well, I actually got that mixup up with the sword one which I tried as `libsword` and it didn't work, then I mistakenly "corrected" both of them. I've fixed my patch so that PR should be right now.

Not having the *.pc file on your system is just a matter of not having the right system dependencies, which isn't really part of npm's job.

**Comment by tobias-klein** [comment]
Thanks, you've got a typo in your latest commit (l**u**bcurl instead of l**i**bcurl). Do you mind one more correction?

**Comment by alerque** [comment]
@tobias-klein Oops. Fixed.

**Comment by tobias-klein** [comment]
@alerque I've just merged your patch to ezra-sword-interface and published a new version 0.6.0 of this module on npmjs. You should be able to update locally by first updating the Ezra Project sources to the latest revision from master and then issuing `npm install` another time. Then you can try to continue with `npm run rebuild-linux` and the steps from the build/release instruction. Please confirm that this issue can be closed after verification on your system.
